jump-core/src/main/java/com/bitdecay/jump/BitWorld.java

```diff
@@ -110,64 +115,116 @@ private void internalStep(final float delta) {
 
 		occupiedSpaces.clear();
 
-		// first, move everything
-		bodies.parallelStream().forEach(body -> {
+		/**
+		 * FIRST, MOVE EVERYTHING
+		 */
+		dynamicBodies.stream().forEach(body -> {
 			if (body.active) {
-				// apply gravity to DYNAMIC bodies
-				if (BodyType.DYNAMIC == body.bodyType) {
-					if (body.gravitational) {
-						body.velocity.add(gravity.scale(delta));
-					}
-				}
-				// then let controller handle the body
-				if (body.controller != null) {
-					body.controller.update(delta, body);
-				}
+				updateDynamics(body, delta);
+				updateInput(body, delta);
+				moveBody(body, delta);
+				updateOccupiedSpaces(body);
+				resetCollisions(body);
+			}
+		});
 
-				// cap body at max speed
-				body.velocity.x = Math.min(Math.abs(body.velocity.x), maxSpeed.x) * (body.velocity.x < 0 ? -1 : 1);
-				body.velocity.y = Math.min(Math.abs(body.velocity.y), maxSpeed.y) * (body.velocity.y < 0 ? -1 : 1);
-
-				// then move all of our non-static bodies
-				if (BodyType.STATIC != body.bodyType) {
-					body.lastAttempt = body.velocity.scale(delta);
-					body.aabb.translate(body.lastAttempt);
-					if (BodyType.KINETIC == body.bodyType) {
-						for (BitBody child : body.children) {
-							/*
-							 * we make sure to move the child just slightly less
-							 * than the parent to guarantee that it still
-							 * collides if nothing else influences it's motion
-							 */
-							BitPoint influence = body.lastAttempt.shrink(MathUtils.FLOAT_PRECISION);
-							child.aabb.translate(influence);
-							// the child did attempt to move this additional amount according to our engine
-							child.lastAttempt.add(influence);
-						}
-						body.children.clear();
-					}
-				}
-				// all bodies are assumed to be not grounded unless a collision happens this step.
-				body.grounded = false;
-				// all bodies assumed to be independent unless a lineage collision happens this step.
-				body.parent = null;
+		kineticBodies.stream().forEach(body -> {
+			if (body.active) {
+				updateKinetics(body);
+				moveBody(body, delta);
+				updateOccupiedSpaces(body);
+				resetCollisions(body);
 			}
 		});
 
-		// resolve collisions for DYNAMIC bodies against Level bodies-
-		bodies.stream().filter(body -> body.active && BodyType.DYNAMIC == body.bodyType).forEach(body -> buildLevelCollisions(body));
-		//		applyPendingResolutions();
-		bodies.stream().filter(body -> body.active && BodyType.KINETIC == body.bodyType).forEach(body -> buildKineticCollections(body));
+		staticBodies.stream().forEach(body -> {
+			if (body.active) {
+				updateOccupiedSpaces(body);
+			}
+		});
+		/**
+		 * END OF MOVING EVERYTHING
+		 */
+
+		/**
+		 * BUILD COLLISIONS
+		 */
+		dynamicBodies.stream().filter(body -> body.active).forEach(body -> {
```

Comment:
Is it more efficient to just put the filter part within the forEach loop like you did on green 121 and green 131?  Which ever way is more efficient is the way you should probably use throughout.